### ISSUE TITLE
Outdated replica in nats cluster

### ISSUE THREAD

**Issue by ogermann** [nats-io/nats-server#3192]
Nats Server Version: 2.8.4

We're currently experiencing an issue in our nats cluster with 5 replicas running in a kubernetes cluster with several nodes in different datacenters.

We used the official helm chart to deploy the cluster: https://github.com/nats-io/k8s/tree/main/helm/charts/nats#jetstream

### Problem
The nats-dev-1 replica is outdated:

```
Information for Stream block created 0001-01-01T00:00:00Z

Configuration:

             Subjects: block.*
     Acknowledgements: true
            Retention: File - Limits
             Replicas: 3
       Discard Policy: Old
     Duplicate Window: 2m0s
     Maximum Messages: unlimited
        Maximum Bytes: unlimited
          Maximum Age: 15d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats-dev
               Leader: nats-dev-2
              Replica: nats-dev-4, current, seen 0.01s ago
              Replica: nats-dev-1, outdated, seen 14h38m35s ago, 111,300 operations behind

State:

             Messages: 64,420
                Bytes: 14 MiB
             FirstSeq: 574,579 @ 2022-05-30T15:15:12 UTC
              LastSeq: 638,998 @ 2022-06-14T15:14:50 UTC
     Active Consumers: 46
```

### Things we tried to solve the problem
- Removing the peer via CLI. Which didn't solve the problem since the problem persists when the same replica steps back in for the same stream.

### Log
Unusual errors from the nats-dev-1 replica log:
```
[8] 2022/06/14 00:23:16.989520 [ERR] RAFT [4KDaJ00g - C-R5F-7G2iiks6] Resource not found: open /data/jetstream/AC2K...UY7S/_js_/C-R5F-7G2iiks6/peers.idx: no such file or directory
[8] 2022/06/14 00:23:16.989554 [WRN] RAFT [4KDaJ00g - C-R5F-7G2iiks6] Error writing peer state file for "C-R5F-7G2iiks6": open /data/jetstream/AC2K...UY7S/_js_/C-R5F-7G2iiks6/peers.idx: no such file or directory
```

### Conclusion
We could probably fix this by deleting the volume of the outdated replica. But since this could possibly be a bug we wanted to report it to you and we're kinda worried about this error as to what could happen in our production environment.
We're also experiencing prometheus metrics that do not seem to be correct, as they show pending messages, which we don't see in the CLI (not sure if this could be related).

**Comment by derekcollison** [comment]
Which server are the log messages from above being generated?
2.8.4 requires the peers file, any reason you would know of on why it is not there?

**Comment by ogermann** [comment]
Posting this just in case there is a relation:
As I was inspecting the replica logs a bit further I also saw a lot of these messages in the nats-dev-2 replica logs:
```
[8] 2022/06/14 18:39:40.994040 [ERR] Received an error looking up message for consumer: store is closed
```
Since the nats-dev-2 is the leader of the problematic block stream I performed a step-down which worked and now the nats-dev-4 is the new leader.
The errors in log of the nats-dev-2 remain though, 4 of these popping up per second.

**Comment by derekcollison** [comment]
It seems like the storage directory for that replica is having some issues.

How many FDs is the system allowed to use?

If all streams and consumers are replicated. You could shutdown that server and move it storage directory away (we should keep it for triage) and restart the server with a clean storage directory. It will self heal itself from other servers.

**Comment by ogermann** [comment]
So after investigating this a little further I found out that after restarting the complete nats cluster the replicas restore the streams as follows:
```
nats-dev-1: [8] 2022/06/14 18:51:36.969240 [INF]   Restored 0 messages for stream 'ABZ...2OP > block'
nats-dev-3: [8] 2022/06/14 18:49:56.443268 [INF]   Restored 64,467 messages for stream 'ABZ...2OP > block'
nats-dev-4: [7] 2022/06/14 18:49:05.690538 [INF]   Restored 64,465 messages for stream 'ABZ...2OP > block'
```
while:
```
Cluster Information:

                 Name: nats-dev
               Leader: nats-dev-4
              Replica: nats-dev-1, outdated, seen 2h5m53s ago, 112,325 operations behind
              Replica: nats-dev-3, current, seen 0.25s ago
```

**Comment by ogermann** [comment]
What do you mean by FDs?

**Comment by derekcollison** [comment]
`ulimit -a`

What is displayed for limit of open files..

**Comment by ogermann** [comment]
`open files                      (-n) 1048576`

**Comment by ogermann** [comment]
> If all streams and consumers are replicated. You could shutdown that server and move it storage directory away (we should keep it for triage) and restart the server with a clean storage directory. It will self heal itself from other servers.

AFAIK this is sadly not possible with kubernetes statefulsets since you cannot shutdown a single specific instance, so our only option would be to shutdown the whole cluster and perform the action.

**Comment by derekcollison** [comment]
I am not too much of an expert on K8S, but seeing if @caleblloyd or @wallyqs have some insight.

Seems there should be a way to just shut down one pod..

**Comment by ogermann** [comment]
Ok, thank you!

I just looked into our production system, and it seems that we have the exact same problem with the outdated streams there too (the replicas marked with the "!"):
```
╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                           Stream Report                                                            │
├───────────────────────┬─────────┬───────────┬───────────┬─────────┬──────┬─────────┬───────────────────────────────────────────────┤
│ Stream                │ Storage │ Consumers │ Messages  │ Bytes   │ Lost │ Deleted │ Replicas                                      │
├───────────────────────┼─────────┼───────────┼───────────┼─────────┼──────┼─────────┼───────────────────────────────────────────────┤
│ xx                    │ File    │ 1         │ 0         │ 0 B     │ 0    │ 0       │ nats-0, nats-1*, nats-2  │
│ xx                    │ Memory  │ 0         │ 0         │ 0 B     │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 0         │ 0         │ 0 B     │ 0    │ 0       │ nats-0, nats-1*, nats-2  │
│ xx                    │ File    │ 1         │ 0         │ 0 B     │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 1         │ 0         │ 0 B     │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 1         │ 0         │ 0 B     │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 1         │ 0         │ 0 B     │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
│ patchUser             │ File    │ 1         │ 0         │ 0 B     │ 0    │ 0       │ nats-0, nats-1!, nats-2* │
│ patchUser             │ File    │ 1         │ 0         │ 0 B     │ 0    │ 0       │ nats-0, nats-1*, nats-2! │
│ xx                    │ File    │ 1         │ 0         │ 0 B     │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 1         │ 0         │ 0 B     │ 0    │ 0       │ nats-0, nats-1*, nats-2  │
│ xx                    │ Memory  │ 0         │ 0         │ 0 B     │ 0    │ 0       │ nats-0, nats-1, nats-2*  │
│ xx                    │ File    │ 1         │ 1         │ 693 B   │ 0    │ 187     │ nats-0, nats-1, nats-2*  │
│ xx                    │ File    │ 11        │ 16        │ 1.5 KiB │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 1         │ 11        │ 2.0 KiB │ 0    │ 1       │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 3         │ 7         │ 2.9 KiB │ 0    │ 0       │ nats-0, nats-1*, nats-2  │
│ xx                    │ File    │ 3         │ 135       │ 15 KiB  │ 0    │ 80      │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 2         │ 863       │ 109 KiB │ 0    │ 0       │ nats-0, nats-1*, nats-2  │
│ xx                    │ File    │ 0         │ 743       │ 210 KiB │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 0         │ 2,079     │ 232 KiB │ 0    │ 0       │ nats-0, nats-1*, nats-2  │
│ xx                    │ File    │ 1         │ 1,113     │ 272 KiB │ 0    │ 0       │ nats-0, nats-1*, nats-2  │
│ xx                    │ File    │ 1         │ 2,232     │ 308 KiB │ 0    │ 0       │ nats-0!, nats-1, nats-2* │
│ xx                    │ File    │ 2         │ 5,164     │ 942 KiB │ 0    │ 58641   │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 1         │ 17,298    │ 1.7 MiB │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 0         │ 5,164     │ 1.8 MiB │ 0    │ 8587    │ nats-0*, nats-1, nats-2  │
│ xx                    │ File    │ 6         │ 64,472    │ 14 MiB  │ 0    │ 0       │ nats-0*, nats-2                 │
│ xx                    │ File    │ 1         │ 3,369,521 │ 796 MiB │ 0    │ 0       │ nats-0*, nats-1, nats-2  │
╰───────────────────────┴─────────┴───────────┴───────────┴─────────┴──────┴─────────┴───────────────────────────────────────────────╯
```

This system runs a `nats server version 2.8.1` with 3 replicas all running in the same datacenter. The kubernetes cluster is the same though and the storage provisioner also (https://github.com/LINBIT/linstor-server). 

There also seem to be 2 versions of a stream called "patchUser":
```
Information for Stream patchUser created 0001-01-01T00:00:00Z

Configuration:

             Subjects: patchUser.*
     Acknowledgements: true
            Retention: File - WorkQueue
             Replicas: 3
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false
     Maximum Messages: unlimited
        Maximum Bytes: unlimited
          Maximum Age: unlimited
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats
               Leader: nats-1
              Replica: nats-0, current, seen 0.24s ago
              Replica: nats-2, outdated, seen 0.24s ago, 1 operation behind

State:

             Messages: 0
                Bytes: 0 B
             FirstSeq: 1
              LastSeq: 0
     Active Consumers: 1
```
```
Information for Stream patchUser created 0001-01-01T00:00:00Z

Configuration:

             Subjects: patchUser.*
     Acknowledgements: true
            Retention: File - WorkQueue
             Replicas: 3
       Discard Policy: Old
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false
     Maximum Messages: unlimited
        Maximum Bytes: unlimited
          Maximum Age: unlimited
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


Cluster Information:

                 Name: nats
               Leader: nats-2
              Replica: nats-1, outdated, seen 0.19s ago, 74 operations behind
              Replica: nats-0, current, seen 0.19s ago

State:

             Messages: 0
                Bytes: 0 B
             FirstSeq: 1
              LastSeq: 0
     Active Consumers: 1
```

All 3 replicas didn't log any unusual errors.

**Comment by derekcollison** [comment]
You should upgrade to 2.8.4 for sure.
Do you have a support contract with us?

**Comment by derekcollison** [comment]
After upgrading if there still remains issues let us know. 

Two approaches are..

1. Ask the stream to have its current leader stepdown.
2. Scale down to R1 and back to R3..

**Comment by ogermann** [comment]
We will upgrade ASAP. No we don't have a support contract whit you.
Thank you for pointing out the approaches.

**Comment by caleblloyd** [comment]
> the storage provisioner also (https://github.com/LINBIT/linstor-server)

Is this high performance storage?  Do you know what max IOPS and Throughput is?

We generally recommend against backing JetStream to file systems with low IOPS or Throughput.  If it is commodity hardware clustered into block storage through Kubernetes that is usually orders of magnitude slower than local storage.

You can always try a Host Mount or Local Storage provisioner and see if that fixes things, if so I would suspect the block storage.

**Comment by ogermann** [comment]
I know it's backed by the fastest storage we could get from our cloud service provider, but I don't know the numbers, I'll try to figure that out though.

**Comment by ogermann** [comment]
We'll try to switch to local storage provisioning asap.

But in the meantime, we tried to restore the cluster by deleting the outdated replica's data and starting it again (we did this sequentially for every replica just to be sure to get clean data). Now we have the following error when we try to do something (only actions that seem to write something like publishing a message or editing a stream to any higher replicas):
```
natsbox-persist-574cb65964-l25jb:~# nats -s nats://nats-dev:4222 --creds /nsc/nkeys/creds/op-dev/dev-bknd/dev-bknd.creds stream edit --replicas=4 currency
Differences (-old +new):
  api.StreamConfig{
  	... // 10 identical fields
  	Storage:  s"File",
  	Discard:  s"Old",
- 	Replicas: 3,
+ 	Replicas: 4,
  	NoAck:    false,
  	Template: "",
  	... // 8 identical fields
  }
? Really edit Stream currency Yes
nats: error: could not edit Stream currency: insufficient resources (10023)
```
There arent any logentries pointing to this action.

According to other issues here, we checked the meta leader, and it seems to be fine in all replicas.

It seems to us like the storage is readonly, but we can write to the storage directly.

**Comment by derekcollison** [comment]
What does `nats server report jetstream` show from a system user's credentials?

The system is saying it can not find resources for the stream at R4.

**Comment by derekcollison** [comment]
Hmm that looks like the credentials are incorrect. Maybe @matthiashanel can chime in here..

Also what does `nats account info` report for the account trying to issue the JetStream commands.

**Comment by ogermann** [comment]
```
natsbox-persist-574cb65964-l25jb:~# nats -s nats://xxxxxx-nats-dev:4222 --creds /nsc/nkeys/creds/op-xxxxxx-dev/xxxxxx-dev-bknd/xxxxxx-dev-bknd.creds account info
Connection Information:

               Client ID: 187
               Client IP: 192.168.193.176
                     RTT: 12.556608ms
       Headers Supported: true
         Maximum Payload: 1.0 MiB
       Connected Cluster: xxxxxx-nats-dev
           Connected URL: nats://xxxxxx-nats-dev:4222
       Connected Address: 192.168.77.155:4222
     Connected Server ID: NAGC6W26MUHRHWVGSGBPRSBAPDYUKB6XBEOZIBQPUB3SF3IUHJRJIE3N
Connected Server Name: xxxxxx-nats-dev-0

JetStream Account Information:

Account Usage:

    Storage: 2.7 GiB
     Memory: 0 B
    Streams: 33 B
Consumers: 44 B

Account Limits:

Tier: Default

      Configuration Requirements:

         Stream Requires Max Bytes Set: false
          Consumer Maximum Ack Pending: Unlimited

      Stream Resource Usage Limits:

                    Memory: 0 B of Unlimited
         Memory Per Stream: Unlimited
                   Storage: 2.7 GiB of Unlimited
        Storage Per Stream: Unlimited
                   Streams: 33 of Unlimited
                 Consumers: 44 of Unlimited
```

**Comment by derekcollison** [comment]
And for the credentials being used for the user account..

What does `nsc describe jwt -f <user.creds>` show?

**Comment by ogermann** [comment]
It is the same account, I just masked it differently sorry.

```
natsbox-persist-574cb65964-l25jb:~# nsc describe jwt -f /nsc/nkeys/creds/op-xxxxxx-dev/xxxxxx-dev-bknd/xxxxxx-dev-bknd.creds
+---------------------------------------------------------------------------------+
|                                      User                                       |
+----------------------+----------------------------------------------------------+
| Name                 | xxxxxx-dev-bknd                                          |
| User ID              | UDAFLMMJSICVAODGJLZSL7VIJGLSZUZ3XN3QR3LTNZUQODAY54334RS6 |
| Issuer ID            | ABZXRPQAQ53PJG7ZRW2VWWYC7FATHPJKFUEBHGV2K3RMMTENUM7U62OP |
| Issued               | 2021-10-16 11:29:48 UTC                                  |
| Expires              |                                                          |
| Bearer Token         | No                                                       |
+----------------------+----------------------------------------------------------+
| Pub Allow            | >                                                        |
| Sub Allow            | >                                                        |
| Response Permissions | Not Set                                                  |
+----------------------+----------------------------------------------------------+
| Max Msg Payload      | Unlimited                                                |
| Max Data             | Unlimited                                                |
| Max Subs             | Unlimited                                                |
| Network Src          | Any                                                      |
| Time                 | Any                                                      |
+----------------------+----------------------------------------------------------+
```

**Comment by derekcollison** [comment]
You don't need the pub and sub allow of ">", that is the default.

**Comment by ogermann** [comment]
Ok, thanks. It wouldn't make any difference though right?

**Comment by derekcollison** [comment]
Probably would need to Zoom to triage further at this point.

**Comment by ogermann** [comment]
We finally managed to fix this problem by recreating the streams via backup/rm/restore every single stream. After restoring the streams the "insufficient resources (10023)" error went away.

We'll still switch to a faster (local) file storage and observe the streams for the initial issue of outstanding peers.

In the meantime, if the issue of the outstanding peers occurs scaling the stream to R1 (via `nats stream edit --replicas=1 <stream-name>`) and back helps.

Closing this issue for now. If it returns with faster storage we'll reopen it.

Thank you very much for your help @derekcollison  and @caleblloyd.
